src/coreclr/inc/corinfo.h

```diff
@@ -344,16 +344,93 @@ enum StructFloatFieldInfoFlags
     STRUCT_FLOAT_FIELD_SECOND     = 0x4,
     STRUCT_FIRST_FIELD_SIZE_IS8   = 0x10,
     STRUCT_SECOND_FIELD_SIZE_IS8  = 0x20,
+};
+
+// Bitfields for FpStructInRegistersInfo::flags
+namespace FpStruct
+{
+    enum class IntKind
+    {
+        Integer,
+        GcRef,
+        GcByRef,
+    };
+
+    enum Flags
+    {
+        // Positions of flags and bitfields
```

Comment:
The arguments and return values of the JIT/EE interface do not have to be tightly packed bit fields. It is fine for them to take a bit more memory if it makes them simpler and easier to understand.

What would be the simplest and easiest to understand representation of the information that needs to be returned by the JIT/EE interface?

Reply:
(You can also make `getFpStructInRegistersInfo` to take a pointer to the value that it should fill in instead of returning to avoid unnecessary copies of the struct.)

Reply:
> What would be the simplest and easiest to understand representation of the information that needs to be returned by the JIT/EE interface here?

`CORINFO_SWIFT_LOWERING` is a good existing example of what it might look like.

Reply:
> The arguments and return values of the JIT/EE interface do not have to be tightly packed bit fields. It is fine for them to take a bit more memory if it makes them simpler and easier to understand.
> 
> What would be the simplest and easiest to understand representation of the information that needs to be returned by the JIT/EE interface here?

In #101796 I actually experimented a bit with different representations but went back to LoongArch's original design, tweaking only the flag names a bit (and expanding the size of course). Having each of the four field permutations in separate flags turned out to be very convenient in use. And having the sizes and the integer field kind represented in separate fields doesn't change improve readability much as we have a getters for them anyway. I would leave it as is if it's not a problem.

> (You can also make getFpStructInRegistersInfo to take a pointer to the value that it should fill in instead of returning to avoid unnecessary copies of the struct.)

Sure, but if the struct fits in two registers, passing by value is better than by ref, I think.

Reply:
This structure does not look like anything else that exists on JIT/EE interface today. I would prefer to maintain the style of the JIT/EE interface. For example, it can be something like this:

```
#define MAX_FP_STRUCT_LOWERED_ELEMENTS 2

struct CORINFO_FP_STRUCT_REG_INFO
{
    CorInfoType loweredElements[MAX_FP_STRUCT_LOWERED_ELEMENTS];
    uint32_t offsets[MAX_FP_STRUCT_LOWERED_ELEMENTS];
    size_t numLoweredElements;
};

// Returns true when the struct is enregisterable into floating point registers.
// CORINFO_FPSTRUCT_INFO contains additional details in that case.
// Returns false when the struct should be passed according to the integer calling convention.
// Only valid on a RiscV64 and LoongArch64. 
bool getFPStructInRegistersInfo(CORINFO_CLASS_HANDLE cls, CORINFO_FP_STRUCT_REG_INFO* pFPStructInfo);
```

Reply:
I note that the `CORINFO_FP_STRUCT_REG_INFO` is nearly identical to the existing `CORINFO_SWIFT_LOWERING` struct. I am wondering whether the lowering information is a common ABI concern, and we should just use a single type for it. It does not have to be in this PR. I am fine with keeping it separate for now.

Reply:
Not a cheap rework. But fine, I can see it opening doors for some re-use in the future.

Reply:
> Not a cheap rework.

I may have underestimated that.. Looks like `CORINFO_SWIFT_LOWERING` is a strictly JIT interface type, it's used nowhere in `ArgIteratorTemplate`, or around `CallDescrWorker`, profiler, etc, like `FpStructInRegistersInfo`. And all the paraphernalia for working with `CorInfoType` like `CorInfoTypeSize` are in the interpreter.

We can try a Swift-like interface (possibly merging with `CORINFO_SWIFT_LOWERING` type) later in a more focused PR, I don't think I'll have the time for that much reworking now. Most of this PR was supposed to be extending `StructFloatFieldInfoFlags` (already familiar to RV/LA devs) with offsets and all sizes, and subsequent PRs stemming from #101796 assume that.

I'm ok with merging `get{LoongArch,RiscV}64PassFpStructInRegistersInfo`, though, and did so.

Reply:
We prefer to use dedicated types for JIT/EE interface and avoid using those same types throughout the rest of the runtime. It allows us to optimize the JIT/EE interface contract for simplicity and optimize the data representation used at runtime differently (compactness/size, throughput, ....). You can find many instances of this pattern on JIT/EE interface.

I would be ok with keeping what you have for ArgIterator, etc. My concern is specifically around the JIT/EE interface design.

Reply:
Using CORINFO_FPSTRUCT_LOWERING in JIT only makes more sense, I was converting FpStruct::Flags to var_types quite often there, might as well put a similar conversion to CorInfoType in JIT/EE interface. I'll should have it ready soon.

Reply:
> I'll should have it ready soon.

ready